### ISSUE TITLE
alpaka buffers do not respect `const` correctness ? 

### ISSUE THREAD

**Issue by fwyzard** [alpaka-group/alpaka#1729]
As we try to use alpaka Buffer objects in the CMS software, we are facing a problem with `const` correctness due to the way the buffers are implemented.

Here is a sample program, compiled with `g++ -std=c++17 -O3 -Wall -I/home/fwyzard/src/alpaka-group/alpaka/include -DALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED test.cc -o test`:
```c++
#include <alpaka/alpaka.hpp>

template <typename TBuf>
void modify_const_buffer(TBuf buffer) {
  // buffer is not const, and can be used to modify the contents of the originally const buffer:
  buffer[42] = 0;
}

int main(void) {
  using Dim = alpaka::DimInt<1u>;
  using Idx = std::size_t;
  using Host = alpaka::AccCpuSerial<Dim, Idx>;

  const auto host = alpaka::getDevByIdx<Host>(0u);

  const auto buffer = alpaka::allocBuf<int, Idx>(host, alpaka::Vec<Dim, Idx>(64u));
  // buffer is const, so it's not possible to write into it:
  /*
test.cc: In function ‘int main()’:
test.cc:22:14: error: assignment of read-only location ‘((const alpaka::internal::ViewAccessOps<alpaka::BufCpu<int, std::integral_constant<long unsigned int, 1>, long unsigned int> >*)(& buffer))->alpaka::internal::ViewAccessOps<alpaka::BufCpu<int, std::integral_constant<long unsigned int, 1>, long unsigned int> >::operator[](42)’
   22 |   buffer[42] = 0;
  */
  //buffer[42] = 0;

  modify_const_buffer(buffer);

  return 0;
}
```

The issue is that the original `buffer` is `const`, but any non-`const` copy of `buffer` can be used to modify the same data.

Is this the expected behaviour ?

**Comment by sbastrakov** [comment]
I believe your report concerns not const correctness, as this one actually holds for buffers and one can't modify elements of a `const` buffer. However, what alpaka does in this example is automatically casting const buffers to non-const ones e.g. here when passing to a function. Doing a similar thing implicitly (without a `const_cast`) with raw pointers would be a compiler warning/error, but not with our buffers.

I am not sure if this behavior is by design or by accident, we should look into it. I am personally inclined to not allowing it (but again there may be some intent there which i am not familiar with)

**Comment by fwyzard** [comment]
Note that this can happen any time a buffer is copied:
```c++
const auto buffer = alpaka::allocBuf<int, Idx>(host, alpaka::Vec<Dim, Idx>(64u));
// does not compile:
// buffer[42] = 0;

// modifies the contents of the const buffer:
auto copy = buffer;
copy[42] = 0;
```

**Comment by sbastrakov** [comment]
Yes, so alpaka buffer copy effectively performs a `const_cast`. Which a user likely does not expect, especially since raw pointers don't work that way (on modernish compilers).

Just to clarify my point, sorry if somewhat repeating. In this last code sample it's *not* a problem that values inside `buffer` are modified externally - this variable being `const` only means that *it* does not do it (and const correctness meaning it cannot do it), not that the buffer is immutable in principle. However, the automatic cast `auto copy = buffer;` arguably should not compile.

**Comment by fwyzard** [comment]
> the automatic cast `auto copy = buffer;` arguably should not compile.

This is not a cast, it's a copy: `copy` and `buffer` are two different objects, that _point_ to the same data.

**Comment by sbastrakov** [comment]
You are right, I was still thinking of the `const_cast` earlier in my message but of course it was a wrong way to describe `auto copy = buffer;`.

But tbh i still not sure my initial point was clear. E.g. in your last code sample (repeating just for clarity):
```cpp
const auto buffer = alpaka::allocBuf<int, Idx>(host, alpaka::Vec<Dim, Idx>(64u));
// does not compile:
// buffer[42] = 0;

// modifies the contents of the const buffer:
auto copy = buffer;
copy[42] = 0;
```
there is nothing wrong and no violation of const correctness here. Variable `buffer` is `const` and so `buffer[42] = 0;` rightfully would not compile, thus the alpaka buffer indeed being const corrrect. Variable `copy` is not `const` and thus allows mofidication of its elements. Whether the memory for elements is shared between `buffer` and `copy` is not a matter of const correctness, but rather the buffer type having value vs. pointer semantics. Alpaka does the pointer semantics here, which we probably should better document. So basically your very first example function template `modify_const_buffer(TBuf buffer)` also says it can modify buffer elements for `TBuf` being an alpaka buffer, similar to a `foo(int *buf)` saying it can modify the elements. If it doesn't mean to modify, `const` should be added to the function argument (or in the template case perhaps the call made such that `TBuf = const alpaka buffer`).

**Comment by fwyzard** [comment]
> So basically your very first example function template `modify_const_buffer(TBuf buffer)` also says it can modify buffer elements for `TBuf` being an alpaka buffer, similar to a `foo(int *buf)` saying it can modify the elements.

No, the problem is that also a function declared like
```c++
template <typename TBuf>
void modify_const_buffer(TBuf const& buffer);
```
can modify the contents of the buffer:
```c++
template <typename TBuf>
void modify_const_buffer(TBuf const& buffer) {
  // making a copy of a const buffer allows modifying the underlying data
  auto copy = buffer;
  copy[42] = 0;
}
```

So, the summary is that alpaka buffers can never give any `const` guarantee.

**Comment by sbastrakov** [comment]
Okay, so there is difference in understanding and formulating this as clear from the VC, i see no point arguing about it further. It indeed does not give a guarantee needed for the use case described which is what eventually matters and not how we prefer to think about it.

From my pov alpaka buffers give const-correctness as much as their pointer semantic allows. In my view, which as we figured on the VC is different from yours, a const alpaka buffer merely means that this variable cannot be used to modify the contents. It does not mean that the contents of the container are immutable in principle. So that applies to `buffer` variable, but not to `copy` in your examples.

Here there is really a diffenence between value-semantic and pointer-semantic containers. For value-semantic containers like `std::vector` (so with deep copies on assignment and copy construction) there cannot be two `std::vector` variables which share the data. So in this case there is no difference in our points of view on const correctness as the `buffer` is the only vector that could allow modifying the elements, and that is `const` so the vector elements are immutable in this context. But for pointer semantics like alpaka buffer here we meant different things by const correctness, hence the argument.

**Comment by sbastrakov** [comment]
Yes, so can this function:
```cpp
template <typename T>
void modify_const_buffer(const T* buffer){
  // making a non-const accessor to data
  auto * copy = const_cast<T*>(buffer);
  copy[42] = 123;
}
```

Imo your last code sample expresses exactly that: its interface says it will not modify the elements, but internally it does by making another non-const accessor to the same data. Of course, with alpaka and templates it is much less clear that it's what happens in your function and very easy to write this unintentionally, here i would definitely agree. As this code has an ugly `const_cast` that also tells I'm doing something potentially dangerous, while in alpaka case it's a mere `=`. So then it can mean that the buffer is very awkward to use for your actual class and some kind of view (which we may need to extend/implement in alpaka as discussed) would be much nicer.

**Comment by fwyzard** [comment]
> From my pov alpaka buffers give const-correctness as much as their pointer semantic allows. In my view, which as we figured on the VC is different from yours, a const alpaka buffer merely means that this variable cannot be used to modify the contents. It does not mean that the contents of the container are immutable in principle. So that applies to `buffer` variable, but not to `copy` in your examples.

Yes, I understand this.

What you are saying is that a `Buf` containing an `int` should be considered like an `int*` or a `shared_ptr<int>`:
  - a `const Buf` is just like a `int * const`: the pointer itself cannot be modified, but it can be copied to a `int *`, and the pointed-to data can always be modified
  - a `const Buf` is _not_ like a `const int*`, that is, a `const` buffer is not like a pointer to `const` object

**Comment by sbastrakov** [comment]
In my opinion it's as follows

> What you are saying is that a Buf containing an int should be considered like an int* or a shared_ptr<int>

For this part (but not the bullet points) semantically like an `int*`, yes. I think standard C++ smart pointers for arrays have issues of their own unrelated to alpaka and so should not be used for that (but that's besides our current discussion). But we can indeed use a raw pointer as a model for alpaka buffer copying and sharing.

>  a `const Buf` is just like a `int * const`: the pointer itself cannot be modified, but it can be copied to a `int *`, and the pointed-to data can always be modified
>   a `const Buf` is *not* like a `const int*`, that is, a `const` buffer is not like a pointer to `const` object

This highlights our difference in views very well. In my opinion alpaka const buffer is actually somewhere in between those two types, and I guess we are rounding off to different sides of it. I tend to think of it more as a `const int * const` with a downside of "downmoting" to just `int * const` in situations like `auto copy = buffer;`. My point is that a standalone variable of type `const Buf` allows neither itself nor the data to be modified, same as a `const int * const` would. (and that's how i understand const correctness for pointer-semantic containers)

Of course, as you pointed out, on the other hand a const buffer allows writing `auto copy = buffer;` (or pass a buffer by value to a function, like in your very first example), and then the `copy` already behaves like an `int *` pointing to the same memory as `buffer`. This would not automatically happen to raw pointers, where a `const_cast` would be required for such a change, or an `-fpermissive` compiler flag. But alpaka allows just that, and potentially worse as an innocuously looking `=`.

So in my view the line `auto copy = buffer;` in all your samples expresses that - telling to effectively remove constness from the buffer (yes, it's not casting technically but feels similar logically) and then of course changes could happen to its elements.  This operation is in principle unsafe, same as `const_cast`, and so this is not what a function taking a `const Buffer` should be doing. So again here current alpaka implementation expects a user to follow that, but does't forbid it or at least provide a function with a meaninful name that does this operation.

**Comment by fwyzard** [comment]
Sorry, but no. This

> I tend to think of it more as a `const int * const` with a downside of "downmoting" to just `int * const` in situations like `auto copy = buffer;`.

is not C++, it's something you are making up yourself.

C++ does not have any concept of "downmoting" a `const T*` to a `T*`. That requires a very explicit `const_cast<T*>`.

**Comment by sbastrakov** [comment]
Yes, so we again arrive that a const alpaka buffer is neither exactly like `int * const` nor exactly like `const int *`. So we never are exaclty like a native C++ type behavior

**Comment by sbastrakov** [comment]
Also if we are being technical `auto copy = buffer;` means to throw away `const` if a (non-pointer) type of `buffer` had it. So technically here a user is asking to remove `const` and that happens. So one could argue this one is somewhat logical, but not a desired behavior here.

**Comment by sbastrakov** [comment]
Okay well i'm off - i think we now understand what's happening and what's a solution and arguing about naming and formulations

**Comment by fwyzard** [comment]
> Also if we are being technical `auto copy = buffer;` explicitly tells to throw away `const`

No, it does not, and that's the whole point !
It means to make _a copy_ of an object, not to give you a way to bypass the fact that the original object is `const` !

> So one could argue this one is somewhat logical, but not a desired behavior here.

It is neither logical, nor consistent with how the vast majority of C++ objects work.

If you say "that's how alpaka objects behave" I obviously agree.
If you say "that's how alpaka objects _should_ behave and how anybody with some experience with C++ would _expect_ them to behave", then I disagree.

**Comment by sbastrakov** [comment]
I'm confused now. When writing `auto copy = buffer;` and `buffer` not being of a pointer or reference types, the cv qualifiers will be dropped. So in this case `auto` tells that type of `copy` is type of `buffer` without cv qualifiers. How is that not telling to drop const?

**Comment by sbastrakov** [comment]
I even agree with your last point. E.g. especially with writing templated code, or being far away from alpaka. So i'm now just puzzled what we are arguing about at all

So yes, i very much agree alapka should not allow this code to work, as it's very easy to write and make a mistake without realizing it. And it can probably be solved simply by having a function with explicit name instead of `operator =`, one for a "safe copy" const to const, and another one for unsafe const to non-const

**Comment by sbastrakov** [comment]
Okay, maybe now i realized what was missing in all my posts in this issue. What i express is of course how alpaka works as of now, and so how a user code should be written for as-is alpaka. Basically, with understanding that  `auto copy = buffer;` is unsafe and generally should be avoided for `const Buf`, with `const auto copy = buffer;` as a safe version. I am not saying alpaka should be working like this, as it's extremely confusing indeed, and I thought i expressed it in most my messages but then probably did it poorly

**Comment by fwyzard** [comment]
Note that I do not have a better solution to suggest :-/

One could make the copy and assignment operators only accept non-`const` buffers as argument - but that's likely to break a lot of code.

**Comment by sbastrakov** [comment]
I guess a classical solution is to have a `ConstBuffer` type with matching interface, so that `const` is part of the type and this is truly a "shallow copy const correct" container in the sense you need. So then instead of `const Buf buffer` a user would use `ConstBuf buffer`, and proper conversion functions between the buffers. Not that i particularly like it, but we could evaluate it in principle. Perhaps that can even done on a user side in principle on top of alpaka buffers? (but it's not easy to evaluate that quickly)

Or alternatively what was mentioned on the call, instead to have some kind of a constant view and make it so that const buffers produce constant views.

**Comment by fwyzard** [comment]
> I guess a classical solution is to have a `ConstBuffer` type with matching interface, so that `const` is part of the type and this is truly a "shallow copy const correct" container in the sense you need.

This could work, together with a change to the `Buffer` copy constructor: a non-`const` Buffer can be used to construct a `Buffer`, while a `const` Buffer can only be used to construct a `ConstBuffer`.


> Or alternatively what was mentioned on the call, instead to have some kind of a constant view and make it so that const buffers produce constant views.

I'm actually considering the other approach we discussed this morning: to make the interface of the object that holds the buffer expose only `const` access to the data contained in the buffer, rather than to the buffer itself.

**Comment by bernhardmgruber** [comment]
I quickly prototyped a `ViewConst` here: https://github.com/alpaka-group/alpaka/pull/1746.

**Comment by fwyzard** [comment]
Just thinking aloud: if we have a `std::shared_ptr<T>`, we can automatically convert it to a `std::shared_ptr<const T>`:
```c++
#include <iostream>
#include <memory>

using Buffer = std::shared_ptr<int>;
using ConstBuffer = std::shared_ptr<const int>;

Buffer make_buffer() {
  auto buffer = std::make_shared<int>(42);
  return buffer;
}

void dump_buffer(ConstBuffer buffer) {
  std::cout << (*buffer) << std::endl;
}

int main(void) {
  auto buffer = make_buffer();
  dump_buffer(buffer);

  return 0;
}
```

With this code we can be reasonably sure that `dump_buffer(...)` will not modify the content of `buffer` (_i.e._, not without using a `static_cast`).

Unfortunately, an alpaka buffer `TBuf<T, TDim, TIdx>` cannot be converted to a  `TBuf<const T, TDim, TIdx>`, so we cannot use the same approach.

(please note, I'm not saying that having an automatic conversion from `TBuf<T>` to `TBuf<const T>` would solve all the `const`-ness problems, but it could help address some of them; we can try to come up with a more concrete case by next week)

**Comment by bernhardmgruber** [comment]
```c++
using Buffer = alpaka::Buf<...>;
using ConstBuffer = alpaka::ViewConst<Buffer>;

Buffer make_buffer() {
  auto buffer = std::make_shared<int>(42);
  return buffer;
}

void dump_buffer(ConstBuffer buffer) {
  std::cout << (*buffer) << std::endl;
}

int main(void) {
  auto buffer = make_buffer();
  dump_buffer(buffer);

  return 0;
}
```
That may already work, because I forgot to mark the ctor of `ViewConst` `explicit` :)

**Comment by AntonReinhard** [comment]
After some playing around, @SimeonEhrig and I concluded that there is not really a better solution than the `ConstView` that already exists. We considered making the copy constructor return a const element, but this is not a functionality that exists in C++. In general, there just doesn't seem to be a way to use/propagate C++'s `const` in a way that makes sense for this application.

For alpaka2, we could consider an implementation like [this](https://godbolt.org/z/G6szrq9Ye), which is very close to what is in alpaka now, or using a template specialization and concepts version with something like a `bool isConst` value template.

**Comment by fwyzard** [comment]
@AntonReinhard your `ConstBuf` looks better than the current `ViewConst`, because it shares the ownership of the underlying memory - while with the `ViewConst` the user has to guarantee that the memory remains valid.

It could also support a construct like
```c++
// allocate a mutable buffer
auto tmp = alpaka::allocBuf<float, int>(dev, size);
// fill the buffer
tmp[42] = 3.1415f;
// make the buffer immutable
ConstBuf<float, Dim1D, int> buf = std::move(tmp);
// tmp is destroyed, buf owns the memory
```

What do you think about implementing the `ConstBuf` approach for the current version of alpaka ?

**Comment by AntonReinhard** [comment]
> What do you think about implementing the `ConstBuf` approach for the current version of alpaka ?

I can give it a try in the next few days 👍
